paasta_tools/marathon_tools.py

```diff
@@ -324,12 +327,44 @@ def get_calculated_constraints(self, service_namespace_config):
         return [[str(val) for val in constraint] for constraint in constraints]
 
     def get_routing_constraints(self, service_namespace_config):
+        """
+        Returns a set of constraints in order to evenly group a marathon
+        application amongst instances of a discovery type.
+        If, for example, a given app's 'discover' key is set to 'region', then this function
+        computes the constraints required to group the app evenly amongst each
+        of the actual 'region' values in the cluster.
+        It does so by querying the value of the region attribute for each slave
+        in the cluster, returning a GROUP_BY constraint where the value is the
+        number of unique regions.
+
+        :param service_namespace_config: the config for this service
+        :returns: a set of constraints for marathon
+        """
         discover_level = service_namespace_config.get_discover()
-        locations = get_mesos_slaves_grouped_by_attribute(
-            attribute=discover_level, blacklist=self.get_deploy_blacklist(),
-            whitelist=self.get_deploy_whitelist())
-
-        routing_constraints = [[discover_level, "GROUP_BY", str(len(locations))]]
+        slaves = get_slaves()
+        if not slaves:
+            raise NoSlavesAvailableError(
+                "No slaves could be found in the cluster."
+            )
+        filtered_slaves = filter_mesos_slaves_by_blacklist(
+            slaves=slaves,
+            blacklist=self.get_deploy_blacklist(),
+            whitelist=self.get_deploy_whitelist(),
+        )
+        if not filtered_slaves:
+            raise NoSlavesAvailableError(
```

Comment:
Is it definitely ok to let this bubble up? I'd guess it is okay but haven't checked.